array: inline `convert` where possible

topolarity opened the pull request:
This improves a common scenario, where someone wants to `push!` a poorly-typed object onto a well-typed Vector.

For example:
```julia
const NT = @NamedTuple{x::Int,y::Any}
foo(v::Vector{NT}, x::Int, @nospecialize(y)) = push!(v, (; x, y))
```

The `(; x, y)` is slightly poorly-typed here. It could have any type for its `.y` field before it is converted inside the `push!` to a NamedTuple with `y::Any`

Without this PR, the dispatch for this `push!` cannot be inferred:
```julia
julia> code_typed(foo, (Vector{NT}, Int, Any))[1]
 CodeInfo(
1 ─ ...
│   %4 = %new(%3, x, y)::NamedTuple{(:x, :y), <:Tuple{Int64, Any}}
│   %5 = Main.push!(v, %4)::Vector{@NamedTuple{x::Int64, y}}
└──      return %5
) => Vector{@NamedTuple{x::Int64, y}}
```

With this PR, the above dynamic call is fully statically resolved and inlined (and therefore `--trim` compatible)

topolarity committed:
array: inline `convert` where possible

This improves a common scenario, where someone wants to `push!` a
poorly-typed object onto a well-typed Vector.

For example:
```julia
foo(v::Vector{Int}, @nospecialize(x::Integer)) = push!(v, x)
```

Without this PR, the dispatch for this `push!` cannot be inferred:
```julia
julia> code_typed(foo, (Vector{Int}, Integer))[1]
 CodeInfo(
1 ─ %1 = Main.push!(v, x)::Vector{Int64}
└──      return %1
) => Vector{Int64}
```

With this PR, the above dynamic call is fully statically resolved
and inlined.

topolarity committed:
namedtuple: improve `convert` for NamedTuple

It is common to accidentally push a NamedTuple of a slightly mis-matched
type onto a Vector of well-typed NamedTuples:
```julia
foo(v::Vector{@NamedTuple{x::Int,y}}, x::Int, @nospecialize(y)) =
    push!(v, (; x, y))
```

Together with the above commit, this function is now fully inferred and
is compatible with `--trim`

topolarity commented:
BTW @aviatesk it's unrelated to this PR really, I'm a bit surprised that we have so many `@assume_effects :noub` here around convert(...) calls that we don't control. Does that effect bit not also apply to callees?

aviatesk commented:
> BTW @aviatesk it's unrelated to this PR really, but I'm a bit surprised that we have so many `@assume_effects :noub` here around convert(...) calls that we don't control. Does that effect bit not also apply to callees?

Are you referring to the `@_noub_meta` annotation attached to `insert!(a::Array{T,1}, i::Integer, item)`?  
I wasn’t the one who added it, but I believe it’s dangerous to apply that annotation to methods that include `convert`, as user-defined types could be provided. Since `insert!(a::Array{T,1}, i::Integer, item::T)` already has `@_noub_meta` applied, I don’t think it’s necessary to add the annotation to `insert!(a::Array{T,1}, i::Integer, item)`.

aviatesk reviewed (commented):
Would we want to add a test case like:
```julia
include("test/compiler/irutils")
...
let src = code_typed1(foo, (Vector{NT},Int,Any))
    @test any(!iscall((src, push!), src.code)
end
```
?

vtjnash commented on base/array.jl:
```diff
@@ -324,9 +324,13 @@ copyto!(dest::Array{T}, src::Array{T}) where {T} = copyto!(dest, 1, src, 1, leng
 function fill!(dest::Array{T}, x) where T
-    xT = x isa T ? x : convert(T, x)::T
+    @inline
+    x = x isa T ? x : convert(T, x)::T
```
The isa branch is probably now inserted by inference dispatch?

topolarity replied:
Not sure why since the dispatch below is with a different name, but do you think something should be done about it?

topolarity replied:
Or do you mean that this guard is obsolete now because of other improvements in inference?

vtjnash replied:
Ah, I didn't notice that was a different function then. The guard gives inference the opportunity to union-split it (or at least the option to ignore the convert) when encountering abstract types

aviatesk replied:
`convert(ConcreteType, x::AbstractType)` may generally have more than 4 methods, in which case the `isa` branch ins't generated by the inliner, e.g.
```julia
julia> methods(convert, (Type{Int},Any))
# 7 methods for generic function "convert" from Base:
 [1] convert(::Type{T}, x::Ptr) where T<:Integer
     @ pointer.jl:23
 [2] convert(::Type{T}, index::CartesianIndex{1}) where T<:Number
     @ Base.IteratorsMD multidimensional.jl:137
 [3] convert(::Type{T}, x::AbstractChar) where T<:Number
     @ char.jl:192
 [4] convert(::Type{T}, x::Base.TwicePrecision) where T<:Number
     @ twiceprecision.jl:271
 [5] convert(::Type{T}, x::T) where T<:Number
     @ number.jl:6
 [6] convert(::Type{T}, x::Number) where T<:Number
     @ number.jl:7
 [7] convert(::Type{T}, x::T) where T
     @ Base.jl:129
```

vtjnash commented on base/namedtuple.jl:
```diff
@@ -179,10 +179,11 @@ nextind(@nospecialize(t::NamedTuple), i::Integer) = Int(i)+1
 function convert(::Type{NamedTuple{names,T}}, nt::NamedTuple{names}) where {names,T<:Tuple}
-    NamedTuple{names,T}(T(nt))::NamedTuple{names,T}
+    NT = NamedTuple{names,T}
+    (@inline NT(nt))::NT
```
Lost T(NT) here?

topolarity replied:
Intentionally - the definition `Tuple(nt) = (nt...,)` unfortunately is not inlinable by our compiler for a mildly-abstract NamedTuple type like we specialize on here (it gets stuck as an `_apply_iterate`)

Also the compiler has a [hardcoded de-optimization](https://github.com/JuliaLang/julia/blob/4cdd864e535b16e928c2eff43e6f1583bd77209d/base/compiler/abstractinterpretation.jl#L2405-L2407) for any calls to `Tuple(...)` - I'm not sure where it comes from but it is a problem for the MWE in the OP

vtjnash replied:
Oh, huh, seems like a strange de-optimization there. Is this equivalent? In theory this change is potentially also more semantically correct, since there are many cases where the Tuple T contains a Type, which means the convert to T does not actually return something of type T

vtjnash replied:
Seems like @JeffBezanson introduced this problem in
 https://github.com/JuliaLang/julia/pull/33779/files#diff-0fcc5b3f1870b66e0adbebb39ee9ad51f60194ee100f5c51e368c572ade9ca34R800-R801. Maybe we need to revisit that somewhat?

topolarity replied:
Hmm, yes that does seem problematic I agree - probably worth fixing up soon

topolarity commented:
> Are you referring to the `@_noub_meta` annotation attached to `insert!(a::Array{T,1}, i::Integer, item)`? I wasn’t the one who added it, but I believe it’s dangerous to apply that annotation to methods that include `convert`, as user-defined types could be provided. Since `insert!(a::Array{T,1}, i::Integer, item::T)` already has `@_noub_meta` applied, I don’t think it’s necessary to add the annotation to `insert!(a::Array{T,1}, i::Integer, item)`.

Thanks for the info - I agree, probably best to lift the `convert` and mark the rest with `:noub`

FWIW I was also referring to the `@_noub_if_noinbounds_meta` in `setindex!`, which I've just added to this PR - I think it's pretty much the same story there

topolarity committed:
Update `setindex!` and add test

aviatesk commented:
> FWIW I was also referring to the `@_noub_if_noinbounds_meta` in `setindex!`, which I've just added to this PR - I think it's pretty much the same story there

I think `@_noub_if_inbounds` in `setindex!` is fine. It’s specifically applied to the use of `@boundscheck` within `setindex!`, indicating that there’s no need to taint `:noub` if the caller context isn’t `@inbounds`.

aviatesk commented:
> What makes it specific to the `@boundscheck`?
> 
> I thought that macros at function entry like this apply to the entire body:

You're right that it applies to the entire method, but the `:noub_if_noinbounds` effects bit is designed to not taint with `:noub` unless the caller context is using `@inbounds` _and_ the methods doesn't have anything that taints `:noub`. If a user-defined `convert` is `!:noub`, then the method calling it will also be tainted as `!:noub`, which takes precedence over `:noub_if_noinbounds`.

topolarity commented:
> You're right that it applies to the entire method, but the `:noub_if_noinbounds` effects bit is designed to not taint with `:noub` unless the caller context is using `@inbounds` _and_ the methods doesn't have anything that taints `:noub`. If a user-defined `convert` is `!:noub`, then the method calling it will also be tainted as `!:noub`, which takes precedence over `:noub_if_noinbounds`.

Ah gotcha, thanks for the explainer - We should maybe update the docs in `effects.jl`, they definitely made me think the opposite

aviatesk commented:
Well, if the method's `effects.noub === NOUB_IF_NOINBOUNDS`, then the
> method is guaranteed to not execute any undefined behavior
> if the caller does not set nor propagate the `@inbounds` context
, so the docstring sounds to be correct to me.

The key point I think is the order of `:noub` effect bit, which is `ALWAYS_TRUE > NOUB_IF_NOINBOUNDS > ALWAYS_FALSE`, where `NOUB_IF_NOINBOUNDS` can be transitioned to `ALWAYS_FALSE`.

topolarity commented:
That doesn't sound the same to me. By my understanding, the docstring says that:
```julia
function foo()
    @_noub_if_noinbounds_meta
    do_ub!() # !:no_ub
end
```
>  is guaranteed to not execute any undefined behavior
if the caller does not set nor propagate the @inbounds context

But that's not true - `do_ub!` executes undefined behavior.

topolarity commented:
OK upon further reflection I think the docs are right, and `:noub_if_noinbounds` is invalid to have in the presence of these `convert(...)` after all - removed the effect for now

topolarity commented on base/array.jl:
```diff
@@ -980,12 +984,22 @@ Dict{String, Int64} with 2 entries:
 function setindex!(A::Array{T}, x, i::Int) where {T}
+    @inline
+    x = x isa T ? x : convert(T, x)::T
+    return _setindex!(A, x, i)
```
@aviatesk I'd like to mark this `_setindex!` for `@propagate_inbounds` without propagating for the `convert(...)` above

Do you know of a way to do that?

aviatesk replied:
It's fine to just add `@propagate_inbounds`, since it won’t incorrectly refine the effects of `setindex!(A::Array{T}, x, i::Int)`. If a caller calls `setindex!(A::Array{T}, x, i::Int)` in an `@inbounds` context, the caller’s `:noub` should be tainted accordingly.

topolarity replied:
Right, the effects propagation is fine, but I'm not sure I agree with the `@boundscheck` semantics

This is equivalent to whether:
```julia
@inbounds v[i] = obj_with_internal_arrays
```

should elide `@boundscheck` code in the implicit `convert`

IMO, it really shouldn't - I may have no idea what logic is in the implicit `convert` but the logic I'm trying to certify is just the `x[i]` lookup

aviatesk replied:
Are you referring to the possibility that user-defined `convert` might use `@boundscheck`? I think that’s pretty unlikely, but if a user is using `@boundscheck`, they should be aware that it could be elided by a `@propagate_inbounds` caller that propagates `@inbounds`.

topolarity replied:
Fair enough - I do agree that would be strange to use `@boundscheck` on a `convert` and that's probably how it's been working so far

This still feels like a mis-design to me - `@propagate_inbounds` is used to propagate an assumption on the inbounded-ness of your arguments to the inbounded-ness of the arguments on a callee.

But a `set/getindex` implementation might do `x[i]` in multiple places and they may not all be an x/i that can legally assume inbounded-ness. Seems like it should really be a call-site annotation (just like normal `@inbounds` - it's just the caller-deferred version)

topolarity replied:
In any case, I agree with you that this is an unlikely issue (convert takes no index arguments) so I'll go ahead and add the propagation

aviatesk commented:
> OK upon further reflection I think the docs are right, and `:noub_if_noinbounds` is invalid to have in the presence of these `convert(...)` after all - removed the effect for now

Right.. I somehow overlooked that. It would be fine if we could add the call-site `Base.@assume_effects :noub_if_noinbounds` to `@boundscheck`, otherwise adding `@_noub_if_noinbounds` to the method definition like in this case, we’ll need to lift `convert`.

topolarity committed:
Add `@propagate_inbounds` to `setindex!`

We don't actually want to apply this to the `convert(...)` call
(strictly speaking it's incorrect), but since that call doesn't
have any indices it's unlikely to have any `@boundscheck` code
that we need to worry about.

topolarity commented:
OK, this should be ready for review again - I think all of the effects-/inbounds-related details have been sorted out

oscardssmith commented on base/array.jl:
````diff
@@ -1745,12 +1771,16 @@ julia> insert!(Any[1:6;], 3, "here")
 ```
 """
 function insert!(a::Array{T,1}, i::Integer, item) where T
+    @inline
````
what is this inlining?

topolarity replied:
This function - it's equivalent to an annotation on the definition (iiuc)

aviatesk commented on base/array.jl:
```diff
@@ -980,12 +984,24 @@ Dict{String, Int64} with 2 entries:
 function setindex! end
 
 function setindex!(A::Array{T}, x, i::Int) where {T}
+    @_propagate_inbounds_meta
+    @inline
```
```suggestion
    @_propagate_inbounds_meta
```
`@_propagate_inbounds_meta` subsumes `Expr(:meta, :inline)`, so no need to add `@inline` here.

aviatesk commented on base/array.jl:
```diff
@@ -980,12 +984,24 @@ Dict{String, Int64} with 2 entries:
     return A
 end
 function setindex!(A::Array{T}, x, i1::Int, i2::Int, I::Int...) where {T}
+    @_propagate_inbounds_meta
+    @inline
```
```suggestion
    @_propagate_inbounds_meta
```

aviatesk commented on base/array.jl:
````diff
@@ -1745,12 +1771,16 @@ julia> insert!(Any[1:6;], 3, "here")
 ```
 """
 function insert!(a::Array{T,1}, i::Integer, item) where T
+    @inline
````
```suggestion
    @_propagate_inbounds_meta
```
since now `_insert!` contains `@inbounds`

topolarity replied:
Good catch 👍

topolarity committed:
Add `@propagate_inbounds` to `insert!`

vtjnash reviewed (commented):
SGTM

giordano commented on base/array.jl:
```diff
@@ -980,12 +984,22 @@ Dict{String, Int64} with 2 entries:
+    x = x isa T ? x : convert(T, x)::T
+    return _setindex!(A, x, i1, i2, I...)
+end
+function _setindex!(A::Array{T}, x, i1::Int, i2::Int, I::Int...) where {T}
```
@topolarity should this function have been changed to
```julia
function _setindex!(A::Array{T}, x::T, i1::Int, i2::Int, I::Int...) where {T}
    @inline
    @_noub_if_noinbounds_meta
    @boundscheck checkbounds(A, i1, i2, I...) # generally _to_linear_index requires bounds checking
    memoryrefset!(memoryrefnew(A.ref, _to_linear_index(A, i1, i2, I...), false), x, :not_atomic, false)
    return A
end
```
to match the `_setindex!` method for linear indices?
